### PR TITLE
Plugins

### PR DESCRIPTION
**Pull request by bsideup**

### CODE REVIEW

File: app/src/main/java/com/github/bsideup/liiklus/Application.java [hunk 1]
```diff
@@ -17,9 +26,102 @@ public static void main(String[] args) throws Exception {
...
+                            for (String libDirectory : this.pluginClasspath.getLibDirectories()) {
+                                Path libPath = jarFileSystem.getPath(libDirectory);
+                                if (Files.exists(libPath)) {
+                                    Files.walk(libPath, 1).filter(it -> !Files.isDirectory(it)).forEach(it -> {
+                                        try {
+                                            Path tempFile = Files.createTempFile(it.getFileName().toString(), ".jar");
+                                            Files.copy(it, tempFile, StandardCopyOption.REPLACE_EXISTING);
+                                            pluginClassLoader.addURL(tempFile.toUri().toURL());
+                                        } catch (Exception e) {
+                                            e.printStackTrace();
```
Comment (lanwen) [inline app/src/main/java/com/github/bsideup/liiklus/Application.java]: meh :|

File: app/src/main/java/com/github/bsideup/liiklus/Application.java [hunk 2]
```diff
@@ -17,9 +26,102 @@ public static void main(String[] args) throws Exception {
...
+                        // Don't
+                    }
+
+                    protected void loadJars(Path pluginPath, PluginClassLoader pluginClassLoader) {
+                        pluginClassLoader.addFile(pluginPath.toFile());
+                        try (FileSystem jarFileSystem = FileSystems.newFileSystem(pluginPath, null)) {
+                            for (String libDirectory : this.pluginClasspath.getLibDirectories()) {
+                                Path libPath = jarFileSystem.getPath(libDirectory);
+                                if (Files.exists(libPath)) {
+                                    Files.walk(libPath, 1).filter(it -> !Files.isDirectory(it)).forEach(it -> {
```
[the remark this answers is not in the document]
Comment (bsideup) [reply app/src/main/java/com/github/bsideup/liiklus/Application.java]: nice catch :)

File: app/src/main/java/com/github/bsideup/liiklus/Application.java (hunk 1 above)
Comment (bsideup) [reply app/src/main/java/com/github/bsideup/liiklus/Application.java]: I remember fixing it o_O anyways, fixed now :)

File: app/src/main/java/com/github/bsideup/liiklus/config/RecordPostProcessorChain.java [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+import lombok.RequiredArgsConstructor;
+import lombok.experimental.FieldDefaults;
+
+import java.util.Collection;
+
+@RequiredArgsConstructor
+@FieldDefaults(makeFinal = true)
+public class RecordPostProcessorChain {
+
+    Collection<RecordPostProcessor> processors;
```
Comment (lanwen) [inline app/src/main/java/com/github/bsideup/liiklus/config/RecordPostProcessorChain.java]: is better to store it as List explicitly indicating that collection is sorted?

Comment (bsideup) [reply app/src/main/java/com/github/bsideup/liiklus/config/RecordPostProcessorChain.java]: no, did it intentionally, see TODO in `getAll`. Chains should sort them explicitly (will be added later)

File: plugins/dynamodb-positions-storage/build.gradle [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+dependencies {
+    compileOnly 'org.projectlombok:lombok'
+    compileOnly 'com.google.auto.service:auto-service'
+
+    compileOnly 'org.springframework.boot:spring-boot-starter'
+    compileOnly 'org.springframework.boot:spring-boot-starter-validation'
+    compileOnly project(":api")
+    compileOnly 'io.projectreactor:reactor-core'
+
+    compile ('com.amazonaws:aws-java-sdk-dynamodb:1.11.192') {
```
Comment (lanwen) [inline plugins/dynamodb-positions-storage/build.gradle]: why not the `1.11.321`?

Comment (bsideup) [reply plugins/dynamodb-positions-storage/build.gradle]: the same version as it used to be, don't want to do the dependencies update during this PR, might cause hidden issues

File: plugins/inmemory-positions-storage/src/main/java/com/github/bsideup/liiklus/inmemory/config/InMemoryConfiguration.java [hunk 5]
```diff
@@ -2,21 +2,25 @@
...
 @GatewayProfile
 @ConditionalOnProperty(value = "storage.positions.type", havingValue = "MEMORY")
-public class InMemoryConfiguration {
+public class InMemoryConfiguration implements LiiklusConfiguration {
 
     @Bean
-    InMemoryPositionsStorage inMemoryPositionsStorage() {
+    PositionsStorage inMemoryPositionsStorage() {
         log.warn("\n" +
                 String.format("%0106d", 0).replace("0", "=") + "\n" +
```
Comment (lanwen) [inline plugins/inmemory-positions-storage/src/main/java/com/github/bsideup/liiklus/inmemory/config/InMemoryConfiguration.java]: what this line do?

Comment (bsideup) [reply plugins/inmemory-positions-storage/src/main/java/com/github/bsideup/liiklus/inmemory/config/InMemoryConfiguration.java]: ===================================================

:D

File: app/src/main/java/com/github/bsideup/liiklus/plugins/LiiklusPluginRepository.java [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+
+@RequiredArgsConstructor
+public class LiiklusPluginRepository implements PluginRepository {
+
+    final LiiklusPluginManager liiklusPluginManager;
+
+    @Override
+    @SneakyThrows
+    public List<Path> getPluginPaths() {
+        PathMatcher[] matchers = Stream.of(liiklusPluginManager.getPluginsPathMatcher().split("/"))
```
Comment (lanwen) [inline app/src/main/java/com/github/bsideup/liiklus/plugins/LiiklusPluginRepository.java]: will it grab `./**/*.jar` for any depth pattern?

File: app/src/main/java/com/github/bsideup/liiklus/plugins/LiiklusPluginLoader.java [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+        super(pluginManager, pluginClasspath);
+    }
+
+    @Override
+    protected void loadClasses(Path pluginPath, PluginClassLoader pluginClassLoader) {
+        // Don't load classes, but add plugin's JAR instead
+        pluginClassLoader.addFile(pluginPath.toFile());
+    }
+
+    protected void loadJars(Path pluginPath, PluginClassLoader pluginClassLoader) {
```
Comment (lanwen) [inline app/src/main/java/com/github/bsideup/liiklus/plugins/LiiklusPluginLoader.java]: is it overridden?

**Pull request merged**